sycl/test-e2e/syclcompat/id_query/id_query.cpp

```diff
@@ -0,0 +1,131 @@
+
+#include <algorithm>
+#include <numeric>
+
+#include <syclcompat/device.hpp>
+#include <syclcompat/launch.hpp>
+
+#include "id_query_fixt.hpp"
+
+void global_id_x_query(int *data) {
+  data[syclcompat::global_id::x()] = syclcompat::global_id::x();
+}
+void global_id_y_query(int *data) {
+  data[syclcompat::global_id::y()] = syclcompat::global_id::y();
+}
+void global_id_z_query(int *data) {
+  data[syclcompat::global_id::z()] = syclcompat::global_id::z();
+}
+
+void test_global_id_query() {
+  std::cout << __PRETTY_FUNCTION__ << std::endl;
+
+  constexpr syclcompat::dim3 grid{4};
+  constexpr syclcompat::dim3 threads{32};
+
+  auto checker = [](std::vector<int> input) {
+    std::vector<int> expected(input.size());
+    std::iota(expected.begin(), expected.end(), 0);
+    assert(std::equal(expected.begin(), expected.end(), input.begin()));
+  };
+  // Check we can query x, y, z components of global_id
+  QueryLauncher<global_id_x_query>({4, 1, 1}, {32, 1, 1}).launch_dim3(checker);
+  QueryLauncher<global_id_y_query>({1, 4, 1}, {1, 32, 1}).launch_dim3(checker);
+  QueryLauncher<global_id_z_query>({1, 1, 4}, {1, 1, 32}).launch_dim3(checker);
+
+  // Check that we can query x component, irrespective of the kernel dimension
+  QueryLauncher<global_id_x_query>({4, 1, 1}, {32, 1, 1})
+      .launch_ndrange<3>(checker);
+  QueryLauncher<global_id_x_query>({4, 1, 1}, {32, 1, 1})
+      .launch_ndrange<2>(checker);
+  QueryLauncher<global_id_x_query>({4, 1, 1}, {32, 1, 1})
+      .launch_ndrange<1>(checker);
+
+  // Check we can query y component for 2D kernel
+  QueryLauncher<global_id_y_query>({1, 4, 1}, {1, 32, 1})
+      .launch_ndrange<2>(checker);
+}
+
+void range_x_query(int *data) {
+  data[syclcompat::global_id::x()] = syclcompat::global_range::x() *
+                                     syclcompat::work_group_range::x() *
+                                     syclcompat::local_range::x();
+}
+
+void test_ranges_query() {
+  std::cout << __PRETTY_FUNCTION__ << std::endl;
+
+  constexpr syclcompat::dim3 grid{4};
+  constexpr syclcompat::dim3 threads{32};
+
+  // global_range::x() * work_group_range::x() * local_range::x();
+  int target = grid.x * threads.x * grid.x * threads.x;
+
+  auto checker = [&](std::vector<int> input) {
+    assert(std::all_of(input.begin(), input.end(),
+                       [=](int a) { return a == target; }));
+  };
+  QueryLauncher<range_x_query>(grid, threads).launch_dim3(checker);
+}
+
+void wgroup_id_x_query(int *data) {
+  data[syclcompat::global_id::x()] = syclcompat::work_group_id::x();
+}
+void local_id_x_query(int *data) {
+  data[syclcompat::global_id::x()] = syclcompat::local_id::x();
+}
+
+void test_ids_query() {
+  std::cout << __PRETTY_FUNCTION__ << std::endl;
+
+  constexpr syclcompat::dim3 grid{4};
+  constexpr syclcompat::dim3 threads{32};
+
+  auto wgroup_checker = [&](std::vector<int> input) {
+    for (int i = 0; i < input.size(); ++i) {
+      if (input[i] != i / threads.x)
```

Comment:
Nit: this could be `assert(input[i] == i / threads.x)`, and same below.
Else we could add an informative print statement showing expected & actual value.

Reply:
The assert was refactored. Seems cleaner, I think adding a print statement would require a larger refactor, since right now it would add very little value. We'd only be able to know the first value that fails, with no information of the other expected output.